File: aisd/aisd/src/main/java/org/example/structure/hashtable/HashTableChainMethod.java
package org.example.structure.hashtable;

import org.example.model.Person;

import java.util.ArrayList;
import java.util.List;

// Хэш таблица методом цепочек
public class HashTableChainMethod {
    private final int size; // Размер хэш-таблицы
    private final List<Person>[] data; // Массив списков, используемый для хранения элементов

    public HashTableChainMethod(int size) {
        this.size = size;
        this.data = new List[size]; // Инициализация массива списков
        for (int i = 0; i < size; i++) {
            this.data[i] = new ArrayList<>(); // Инициализация каждого списка в массиве
        }
    }

    public void print() {
        int i = 1;
        System.out.println("======================");
        for (List<Person> element : data) {
            if (element.isEmpty()) {
                System.out.println("Элемент " + i + " - null"); // Вывод информации о пустых элементах
            } else {
                System.out.print("Элемент " + i + " - ");
                for (Person item : element) {
                    System.out.print("| " + item.getName() + " " + item.getAge() + " |"); // Вывод информации о каждом элементе в списке
                }
                System.out.println();
            }
            i++;
        }
        System.out.println("======================");
    }

    public void insert(Person personModel) {
        int index = getIndex(personModel); // Получение индекса для вставки элемента

        if (data[index] == null) {
            data[index] = new ArrayList<>(); // Создание нового списка, если в указанной ячейке еще нет списка
            data[index].add(personModel); // Добавление элемента в список
            return;
        }

        data[index].add(personModel); // Добавление элемента в существующий список
    }

    public void delete(Person personModel) {
        int index = getIndex(personModel); // Получение индекса элемента для удаления

        if (data[index].size() == 1) {
            data[index] = null; // Удаление списка, если он содержит только один элемент
            return;
        }

        for (Person item : data[index]) {
            if (item.getAge() == personModel.getAge() && item.getName().equals(personModel.getName())) {
                data[index].remove(item); // Удаление элемента из списка по критериям
                return;
            }
        }
    }

    public boolean find(Person personModel) {
        int index = getIndex(personModel); // Получение индекса элемента для поиска

        if (data[index] == null) {
            return false; // Возвращение false, если список пуст
        } else {
            for (Person item : data[index]) {
                if (item.getAge() == personModel.getAge() && item.getName().equals(personModel.getName())) {
                    return true; // Возвращение true, если элемент найден в списке
                }
            }
            return false; // Возвращение false, если элемент не найден в списке
        }
    }

    private int getIndex(Person personModel) {
        // Вычисление хэша для определения индекса элемента в массиве
        return (personModel.getAge() * personModel.getAge() + personModel.getName().length() * personModel.getName().length() + personModel.getAge() * personModel.getName().length()) % size;
    }
}

File: misc/gos/ASD/ASD/HashTableChainMethod.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ASD;
public class HashTableChainMethod
{
    private readonly int _size;
    private readonly List<PersonModel>?[] _data;
    public HashTableChainMethod(int size)
    {
        _size = size;
        _data = new List<PersonModel>?[_size];
    }

    public void Print()
    {
        int i = 1;
        foreach (var element in _data)
        {
            if (element == null)
            {
                Console.WriteLine($"Элемент {i} - null");
            }
            else
            {
                Console.Write($"Элемент {i} -");
                foreach (var item in element)
                {
                    Console.Write($"| {item.Name} {item.Age} |");
                }
            Console.WriteLine();
            }
            i++;
        }
    }

    public void Insert(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if (_data[index] == null)
        {
            _data[index] = new List<PersonModel> { personModel };
            return;
        }

        _data[index].Add(personModel);
    }

    public void Delete(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if(_data[index].Count == 1)
        {
            _data[index] = null;
            return;
        }

        foreach (var item in _data[index])
        {
            if (item.Age == personModel.Age && item.Name == personModel.Name)
            {
                _data[index].Remove(item);
                return;
            }
        }
    }

    public bool Find(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if (_data[index] == null)
        {
            return false;
        }
        else
        {
            foreach (var item in _data[index])
            {
                if (item.Age == personModel.Age && item.Name == personModel.Name)
                {
                    return true;
                }
            }

            return false;
        }
    }

    private int GetIndex(PersonModel personModel)
        => (personModel.Age * personModel.Age + personModel.Name.Length * personModel.Name.Length + personModel.Age * personModel.Name.Length) % _size;

}
